applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,888 @@
+void Define3DWakeProcess::SaveLocalWakeNormalInElements() const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    ModelPart& wake_sub_model_part = root_model_part.GetSubModelPart("wake_elements_model_part");
+
+    for (auto& r_elem : wake_sub_model_part.Elements()){
+        NodeType::Pointer p_closest_te_node = *mrTrailingEdgeModelPart.NodesBegin().base();
+        FindClosestTrailingEdgeNode(p_closest_te_node, r_elem.GetGeometry().Center());
+
+        r_elem.GetValue(WAKE_NORMAL) = p_closest_te_node->GetValue(WAKE_NORMAL);
+    }
+}
+
+void Define3DWakeProcess::AddWakeNodesToWakeModelPart() const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    ModelPart& wake_sub_model_part =
+            root_model_part.GetSubModelPart("wake_elements_model_part");
+
+    std::vector<std::size_t> wake_nodes_ordered_ids;
+    for (auto& r_element : wake_sub_model_part.Elements()){
+        for (unsigned int i = 0; i < r_element.GetGeometry().size(); i++){
+            r_element.GetGeometry()[i].SetValue(WAKE, true);
+            wake_nodes_ordered_ids.push_back(r_element.GetGeometry()[i].Id());
+        }
+    }
+
+    std::sort(wake_nodes_ordered_ids.begin(),
+              wake_nodes_ordered_ids.end());
+    wake_sub_model_part.AddNodes(wake_nodes_ordered_ids);
+}
+
+// This function counts the number of elements of each type. Useful for
+// debugging purposes.
+void Define3DWakeProcess::CountElementsNumber() const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    ModelPart& trailing_edge_sub_model_part =
+        root_model_part.GetSubModelPart("trailing_edge_elements_model_part");
+
+    // Initialize counters
+    unsigned int normal_elements_counter = 0;
+    unsigned int wake_elements_counter = 0;
+    unsigned int kutta_elements_counter = 0;
+    unsigned int structure_elements_counter = 0;
+    for (auto& r_element : trailing_edge_sub_model_part.Elements()){
+        if(!r_element.GetValue(WAKE)){
+            if(r_element.GetValue(KUTTA)){
+                kutta_elements_counter += 1;
+            }
+            else{
+                normal_elements_counter += 1;
+            }
+        }
+        else{
+            wake_elements_counter += 1;
+            if(r_element.Is(STRUCTURE)){
+                structure_elements_counter += 1;
+            }
+        }
+    }
+
+    ModelPart& wake_sub_model_part = root_model_part.GetSubModelPart("wake_elements_model_part");
+    const unsigned int all_wake_elements_counter = wake_sub_model_part.NumberOfElements();
+    KRATOS_WATCH(normal_elements_counter)
+    KRATOS_WATCH(kutta_elements_counter)
+    KRATOS_WATCH(wake_elements_counter)
+    KRATOS_WATCH(structure_elements_counter)
+    KRATOS_WATCH(all_wake_elements_counter)
+}
+
+// This function prints the elements ids into a file. Useful for debugging
+// purposes.
+void Define3DWakeProcess::WriteElementIdsToFile() const
```

Comment:
you should only open the files once and `close` them at the end (otherwise esp under win  this can be problematic)

BTW also consider moving this to a utility, doesn't really belong to this already large class IMO

Reply:
Ok!